### ISSUE TITLE
How to pause Silphid.Sequencit

### ISSUE THREAD

**Issue by takaaptech** [Silphid/Silphid.Unity#1]
Hi,
I open sequencing1 demo and found it awesome! Very Clean and polish code!
I am planning to use Sequence and Parallel to manage Wave of enemies, so when user press pause game button, how can I pause the sequence and the parallel actions that made by Sequencit?
In the demo scene, if we have pause and resume button for the demo would be great!
Thank so much!

**Comment by Silphid** [comment]
Hello takaaptech,

Glad to see that you like Sequencit!  I indeed worked very hard to make the syntax as simple and clean as possible.

How to pause sequencing is indeed a very good question. And not a trivial one either.  The pausing must be performed at multiple levels : 1. The Sequence and Parallel classes, 2. The Observables queued within them, and 3. The actions/animations/tweens within the Observables. That is no simple task.

The pausing of #1 and #2 could be achieved by inserting some sort of gates into the sequence, kind of pauseable checkpoints in the sequence (like this: https://stackoverflow.com/a/7642198/959896).  But it would only allow pausing at those specific checkpoints, not at arbitrary moments in time, and it would require to insert those gates between all sequence items where a pause should be allowed to happen. A better approach that would work more globally might be Schedulers. Rx uses Schedulers to schedule actions to be invoked now or later, so a possible solution might be to create a "PauseableScheduler" that could suspend all scheduled actions.

The pausing of #3 could be handled in the extension methods that convert the Tween into an Observable. The extension methods would need to have access to the PauseableScheduler or any pausing state change, so that they may pause the tweens during pauses.

As you see, this is no simple matter.  I never had to achieve that, so it was not built into Sequencit, but now that you mention it, I will keep that in mind and try to think about the best way to approach the problem.

Best regards,
Mathieu

**Comment by Silphid** [comment]
In the "0.99" branch I am working on, I just added support for gates (which handle case #1 above).  If you checkout that branch, have a look at the gate unit tests in Sources/Silphid.Sequencit.Test/Sources/GateTest.cs.

**Comment by Silphid** [comment]
Hello takaaptech,

If you are new to Rx, I recommend that you first familiarize yourself with it.  There are plenty of resources on the Net.  Here are my personal recommendations:

- [The introduction to Reactive Programming you've been missing](https://gist.github.com/staltz/868e7e9bc2a7b8c1f754) - A great intro and overview. The place to start.
- [IntroToRx.com](introtorx.com) - The complete online version of a great book by Lee Campbell, very thorough and insightful.  But start with the previous link first! ;)
- [ReactiveX.io](reactivex.io) - The official reference hub for implementations of Rx on all platforms.
- [RxMarbles](rxmarbles.com) - Interactive diagrams to experiment with and better understand observables and their various operators.
- [UniRx](https://github.com/neuecc/UniRx) - Obviously, the UniRx documentation is another good read, with many good examples specific to Unity.

The thing to keep in mind when working with Sequencit is that you should create all your building blocks as IObservable<Unit>'s (or any other type that implements that interface, such as Sequence, Parallel and Lapse).  The "Unit" type is used in UniRx when you don't care about transferring data in your observable and only want to use the observable to track the completion of something.  For example, you could create a method that creates and tracks one wave of enemies:

private IObservable<Unit> SpawnWave(int numberOfEnemies, float difficultyLevel)
{
    // ...
}

The observable returned by that method allows to track when the wave has been completed (that is, when the observable completes, it means all enemies in the wave have been killed).

Then you could sequence multiple waves like so:

private IObservable<Unit> SpawnAllWaves()
{
    return Sequence.Create(seq =>
    {
        seq.Add(() => SpawnWave(10, 1.0f));
        seq.Add(() => SpawnWave(15, 1.5f));
        seq.Add(() => SpawnWave(20, 2.5f));
    });
}

Now, each enemy's life-time could also be tracked by an IObservable<Unit>. Sequencit provides a simple class for that, named "Lapse" (make sure to pull the latest version of the "0.99" branch, because I just updated it). A "Lapse" is an IObservable<Unit> that completes when you dispose it. So you could dispose it when the enemy is killed (when its game object is destroyed):

public GameObject EnemyPrefab;

private IObservable<Unit> SpawnEnemy(float delayInSeconds, float difficultyLevel)
{
    return Sequence.Create(seq =>
    {
        // So that enemies are spawned after different delays
        seq.AddInterval(delayInSeconds);

        seq.AddLapse(disposable =>
        {
            // Note: This lambda is invoked only once we reach that step in the sequence

            // Spawn enemy game object
            var enemy = EnemyPrefab.Instantiate();

            // TODO: Adjust enemy's hit points according to difficulty level

            // This attaches the lapse's disposable to the enemy's game object.
            // When the game object is destroyed (the enemy is killed), it automatically
            // disposes the lapse (and it will complete).
            disposable.AddTo(enemy);
        });
    });
}

Now, we could implement the SpawnWave() method with a Parallel (which is also an IObservable<Unit>):

private IObservable<Unit> SpawnWave(int numberOfEnemies, float difficultyLevel)
{
    return Parallel.Create(para =>
    {
        for (int i = 0; i < numberOfEnemies; i++)
        {
            var delayInSeconds = i * 2f; // TODO: This could be randomized
            para.Add(SpawnEnemy(delayInSeconds, difficultyLevel));
        }
    });
}

Finally, don't forget to subscribe to your resulting observable, because otherwise nothing will happen. It is the subscription that triggers everything:

var disposable = SpawnAllWaves().Subscribe();

If you want to detect when the player completed all waves you can do it the standard Rx way:

var disposable = SpawnAllWaves().Subscribe(_ => {}, () => Debug.Log("All waves completed!"));

Or using Sequencit way:

var disposable = Sequence.Start(seq => {
    seq.Add(SpawnAllWave);
    seq.AddAction(() => Debug.Log("All waves completed!"));
});

Note the difference between Sequence.Create() and Sequence.Start().  The first one, Create(), only creates the observable and returns it; you can then subscribe to it or composite it with other observables.  The second one, Start(), creates the observable, subscribes to it (which triggers the sequencing of everything it contains) and returns the subscription's disposable.

I have not tested any of the above code, but it should be close to reality.

Good luck! :)
Mathieu

**Comment by Silphid** [comment]
And, by the way, I just had an idea about your initial "pausing" question.  I checked with UniRx and, since UniRx 4.6, the default scheduler (`MainThreadScheduler`) will now respect the current time scale.  That means that, if you pause your game using the usual way, that is, setting the time scale to 0:

``` c#
Time.timeScale = 0;
```

...then any current sequencing that relies on time (like `Sequence.AddInterval()`) should also be paused, for free! :)

This is the best way to pause your game in Unity, because all other animations, particle effects, etc, will also be paused.  And there are ways to have time-scale independent animations/particle-effects/tweens for your menus and other stuff that should not be paused when time scale is 0.

I did not test that, but there's no reason it wouldn't work.

Hope this helps! :)
Mathieu
